### ISSUE TITLE
Can hs.drawing.text make use of ANSI color escapes?

### ISSUE THREAD

**Issue by chdiza** [Hammerspoon/hammerspoon#597]
If HS is to ever really replace GeekTool for me, it will have to be able to display the output of shell scripts (or ruby scripts, etc.) like GeekTool does.  GeekTool will honor color escapes that are emitted by the shell script, e.g.,
```sh
yellow="\033[33m"
none="\033[0m"
echo -e "${yellow}This output${none} is colored"
```
If that script is in a "shell geeklet", there will be yellow text followed by text of whatever text color selection one makes in GeekTool's color picker.

I can't seem to get `hs.drawing.text` to honor these color escapes.  HS prints  `[33mThis output...`

If it makes any difference, I seem to have to use `hs.execute` to first get the shell script's output, then `hs.drawing.text` to put it on the screen.

**Comment by asmagill** [comment]
The problem is that as far as I have seen so far, there is no direct support for ansi sequence to NSAttributedString conversion... and at present, I'm not sure if NSAttributedString with multiple regions (which would be necessary if using that class) would break hs.drawing or not... it would certainly make the text methods more complex.

I'd be curious to know what container class GeekTool is using for its text output... I chose NSAttributedString because its a subclass of NSString and required few changes to the core of hs.drawing... by accepting the limitation of only manipulating the object's entire region and not allow subregions.

Is GeekTool opensource?  If so, it's not out of the question that someone could see how they're handling this and create a new module around that model and not have to worry as much about breaking hs.drawing then...

While I've used Geektool some, I've had issues with it -- probably my fault, as I know people that love and swear by it -- so migrating/replicating things like it's color flexibility has been a low priority for me since I can get most of what I used to use it for already.


I guess the first thing to find out is if we can get at its source... that would go a long way towards determining how much effort a true replacement would take.

**Comment by cmsj** [comment]
@asmagill GeekTool has a long and complex history with being open source, but it has previously had source releases and there are forks still hanging around:

https://github.com/balthamos/geektool-3/blob/ac91b2d03c4f6002b007695f5b0ce73514eb291f/NerdTool/classes/ANSIEscapeHelper.m

**Comment by chdiza** [comment]
Unfortunately I'm unable to help code, but I am more than happy to test whatever anyone comes up with.

HS has an opportunity to gain some converts.  GeekTool has been decaying slowly and slightly with each new OS X release.  But it's seriously buggy now on ElCap, bordering on unusable unless one's geeklets are already created and adjusted they way one wants.

**Comment by asmagill** [comment]
I'll take a crack at it later this week... I have a partial module I've played with off and on which breaks text out into a separate module from drawing, much like we do with images and hs.image, and I think it may lend itself to this... not sure at present how much style control you'll have *after* adding the "Geeklets" output tot he contianer, but I think I can make the ANSI sequences apply on top of the containers already existing formatting style...

Are there sequences I should be aware of beyond the ANSI color ones (and I assume we want both the standard 16 as well as the extended 256 color support)?

**Comment by chdiza** [comment]
> Are there sequences I should be aware of beyond the ANSI color ones (and I assume we want both the standard 16 as well as the extended 256 color support)?

I don't know about non-color escape sequences that we'd want.

I don't think 256 is necessary, and personally, 16 is fine with me.  GeekTool won't do 256, AFAIK.  I guess if it's no harder to do 256 than 16 it couldn't hurt.

**Comment by chdiza** [comment]
I just confirmed that GeekTool doesn't recognize the codes for 256 colors.

**Comment by cmsj** [comment]
@asmagill a separate text module sounds interesting. It is basically just processing text into a userdata object wrapping NSAttributedString?

**Comment by lowne** [comment]
fwiw, if [this](http://indiestack.com/2015/09/presenting-html-on-apple-tv/) works, it might be a more general way to get styled text - NSAttributedString from good old HTML (and it opens up a lot of convenient functionality beyond colouring).
A simple `string.gsub()` should then be sufficient to do ANSi->HTML.

**Comment by lowne** [comment]
Alternatively/additionally, and more on topic with OP's request, ANSI->HTML could be [done form the shell](http://pablomarin-garcia.blogspot.it/2011/04/converting-ansi-to-html-how-to-convert.html) and composed into an `hs.webview`.

**Comment by asmagill** [comment]
Actually the NSAttributedString can be represented fully by a table in Lua... no need for a userdata object.  And I think I've almost solved the problems I was having with multiple attribute ranges within a single string.

The benefit is that most of the conversion for ANSI terminal sequences will be done in Lua, rather than C, and it should be fairly straightforward to extend it to allow other conversions all from Lua as well.

On the backend, if you want your module to support NSAttributedString, you just call [[LuaSkin shared] tableAtIndex:idx toClass:"NSAttributedString"] -- if it's actually a table at index idx, it's parsed out as an NSAttributed string; if instead it's just a string, then an attributed string with the system defaults is returned, so it should be fairly transparent an addition to the programmer at the Lua side of things.

**Comment by cmsj** [comment]
@lowne hey that HTML->NSAttributedString thing looks excellent!

**Comment by asmagill** [comment]
Should be easy enough to add to what I'm working on and would give another avenue for manipulating attributed strings... I wonder if there is a way to go the other direction (NSAttributedString to HTML... there are some HTML "suggestions" which can be put into NSAttributedStrings, so I suspect so... will have to look into it).

In either case, we should be able to write our own conversions for embedded encodings like ANSI in Lua to something else (NSAttributedStirng, HTML for this or hs.webview) shortly

**Comment by asmagill** [comment]
fwiw, in the pull I hope to add shortly, the actual ANSI conversion is done in Lua, so it could easily be extended... it currently supports only the 16 colors (foreground and background), but I think I've commented where changes would be necessary to add support for more colors... most of the other SGR sequences (beyond bold and italic) didn't look worth adding, but it should be easy enough if there is interest.

**Comment by asmagill** [comment]
@chdiza, this will be in the next release... let me know how it works for you... I'm finally able to replace my use of Geektool, though to be honest, I wasn't using it for much, so I'd appreciate hearing from others how well it performs.

**Comment by asmagill** [comment]
(oh, and if you haven't been following it, check https://github.com/Hammerspoon/hammerspoon/pull/620 to see an example of how to use the ansi constructor)

**Comment by chdiza** [comment]
@asmagill I won't have a spare second to test this out until Monday, but I'm looking forward to it.

**Comment by chdiza** [comment]
Actually, I could make some initial tests today.  It mainly works, except for two things:

1. All text displayed on the screen is *extremely* anti-aliased, so much so that it's hard to read and is indeed unreadable at smaller sizes.  This probably has to do with me not having a Retina monitor, but GeekTool doesn't do that and neither do other apps like TextEdit.  (Arguably, Monaco in particular at sizes 10 and 9 shouldn't be antialiased *at all*, much less an overdone antialiasing).

2. The ANSI coloring chokes, especially if some of the to-be-colored text is not ASCII.  By "choke" I mean that the number of characters that are colored doesn't match the number of characters that should be colored.  This little bash script illustrates:
```sh
#!/bin/bash

RED="\033[31m"
NONE="\033[0m"

# Only the digit "9" and the degree symbol should be red.  But the
# first hyphen after the degree symbol is also red.

echo -e "It is ${RED}9°${NONE}---nine!---outside."

# If you change the degree sign to "d", there is still miscoloring; now the "d" and the first hyphen are red, but the "9" is not:
echo -e "It is ${RED}9d${NONE}---nine!---outside."

# The "-e" on echo is necessary, or else HS displays the escape chars themselves.
# If this script is run in the shell, the "-e" is not necessary, and if you 
# have it, the "-e" is printed on the screen.
```

**Comment by asmagill** [comment]
The first I don't see on my non-retina display, but I'll look closer at the GeekTool code and see what font/attribute settings it may be putting into effect that we're not...

The second is because of the 2-byte UTF8 degree symbol... swap their order and the 9d is colored properly, but the 9° is still off... gonna have to think about that one since Lua's indexing for strings is by byte, and Objective-C's (i.e. applying the attribute) is by displayed character.

Keep the tests coming in as you can!

**Comment by chdiza** [comment]
> Keep the tests coming in as you can!

Very well, then!  There seems to be a serious memory leak, but it might be in `hs.timer` rather than the new styledtext.

To reproduce, make the below your `init.lua`.  Then activate HS and follow its mem usage in Activity Monitor.  I see it go up about 8MB of RAM, then drop by 4MB, then go up by 8, down by 4, and it just keeps creeping up like that.  It only does this if that timer is created; the refresh interval governs how fast the mem usage goes up.  I got all the way up to several hundred MB before I gave up.

It's also perfectly possible that I'm being an idiot in this init file :)

```lua
local hotkey   = require "hs.hotkey"
local alert    = require "hs.alert"
local draw    = require "hs.drawing"
local timer   = require "hs.timer"
local geom    = require "hs.geometry"
local rectify = geom.rect

alert.show("Hammerspoon config load attempt", .5)

local my_ls_spot = rectify(100,100,1000,1000)


function run_my_ls()
  return hs.styledtext.ansi(hs.execute("ls"), {font={name="Monaco",size=10},backgroundColor={alpha=1}} )
end

function refresh_my_ls()
  if my_ls then
    my_ls:delete()
  end
  my_ls = hs.drawing.text(my_ls_spot, run_my_ls())
  my_ls:sendToBack()
  my_ls:show()
end


my_ls = nil
refresh_my_ls()

timer_my_ls = hs.timer.new(5, refresh_my_ls, true):start()

local chord    = {"cmd","ctrl","alt"}
local bind     = hotkey.bind
bind(chord,    "0",     function() app.launchOrFocus("iTerm") end)
bind(chord, "F9", function() hs.openConsole() end)
bind(chord, "F8", function() hs.reload() end)
bind(chord, "F7", function() hs.focus(); alert.show("Hammerspoon is focused.") end)

alert.show("Hammerspoon config load complete", .5)
```

**Comment by chdiza** [comment]
Re the over-antialiased text: in the above sample init file, the text looks OK.

If I use Monaco 9 with a transparent background (by not specifying an alpha), and that drawing object is lain on top of a second drawing object that's a black rectangle, *then* the text looks grossly antialiased.  This will be a problem for me, since I need black rectangular backgrounds, separate from the drawing.text object.

**Comment by asmagill** [comment]
My initial thoughts are that it's probably in hs.timer because of the following:

~~~lua
> collectgarbage() ; print(collectgarbage("count")) ; b:delete() ; c = hs.styledtext.ansi(hs.execute("~/Desktop/test3.sh"), {font={name="Monaco",size=10}, color={list="Apple",name="White"}, backgroundColor = {alpha=.5}}) ; b = hs.drawing.text({x=100,y=100,h=42,w=170},c):show() ; print(collectgarbage("count"))
3710.263671875
3726.11328125
~~~

I'm consistently seeing between 10 - 20k, not MB of difference when executing this, though I haven't looked in the activity monitor yet.  I guess its time to take a closer look at hs.timer again as well! :-) 

BTW I do have a fix for the unicode multi-byte issue you pointed out, but its in the Lua code... I'm creating a mapping between the character position Lua thinks is right and converting it to what NSAttributedString thinks is right and applying that to the methods... I'm thinking that the solution really belongs in the Objective-C side, since we're extending Lua and not the other way around... The end user (you) shouldn't have to think about requirements of the Objective-C side when you're coding in Lua. I may go ahead and push this hack since I think initially most of the attention will be on the ANSI convertor and not on creating your own styled text objects through new converters/lua code, but let me see how hard the "correct" solution is when I get a chance to look at it again tonight.

**Comment by asmagill** [comment]
Looking at the NerdTool repository version of GeekTool, I'm not seeing anything obvious that they're doing that we're not (or couldn't)... that said, I don't know how far (if at all) out of sync that repo is with the official GeekTool, so that might not mean anything...

By allowing an hs.drawing object (in this case text, but since its applied at the NSView level, it could in theory be any) to optionally choose to be a layer-backed view, I can see the following difference (these are Monaco at 10pt... I forgot what size you mentioned when I generated the image):

![screen shot 2015-11-01 at 12 32 05 am](https://cloud.githubusercontent.com/assets/8139480/10867471/e7bdc88a-8031-11e5-8829-d65ae2fa41e7.png)

The left version is the default hs.drawing; the right is with the wantsLayers attribute set to yes, and they were both generated with no background and placed on top of a black hs.drawing.rectangle object.  Is that the kind of visual difference you're talking about?

I'm not sure what other changes the layer toggle might introduce or allow us to do, but I can add it to hs.drawing if you think it will help... since it's false by default (which is what it's always been before) there should be no impact on anything currently in use, and who knows, it may suggest some other additions that can be made later.  It does appear that it alters the text height a little, so I may take another crack at hs.drawing.getTextDrawingSize... one of the classes in the NerdTool stuff looks like it may provide a better way to judge size, so I want to take a closer look at that later next week anyways.

**Comment by chdiza** [comment]
It is the *kind* of difference I'm talking about; in that particular screenshot, the *degree* of difference is less than the degree I'm seeing, but that's the basic idea.

If I get time today I'll put up a side-by-side pic.

**Comment by chdiza** [comment]
Re the memory leak, I thought I'd add a couple of notes in case they're relevant.

1. I made the script in question be `ls` just to keep things simple, and to prove that the cause of the leak was not the specific script called.

2. The pattern of "up approx 8MB, down approx 4, up approx 8..." was constant no matter how many separate timers I created.  With one, two, three, or even five timers active, that was the pattern.  Of course, this is on Activity Monitor, which only has tenths of a MB, and it was set to update every second.

**Comment by chdiza** [comment]
> If I use Monaco 9 with a transparent background (by not specifying an alpha), and that drawing object is lain on top of a second drawing object that's a black rectangle, then the text looks grossly antialiased. 

I wrote the above.  I hope I didn't convey the impression that it's *only* when laid upon a separate black rectangle that it looks bad.  A text drawing object with a transparent background looks bad over here either way.  For it to not look bad, I have to make the background not be transparent.

**Comment by asmagill** [comment]
From my research, there are very few things that can be done about jagged edges on small fonts when the background is transparent because of the way OS X rasterizes the fonts as they're displayed... however, if GeekTool is doing something we're not that mitigates this, it would be nice to figure out what.

**Comment by chdiza** [comment]
> however, if GeekTool is doing something we're not that mitigates this

I don't think there is anything they do; I make my text geeklets have a non-transparent background at least in part to work around what you say OS X does.

But GeekTool at least *used* to block the overantialiasing from a different source, namely that which comes from Retina-optimization combined with non-Retina monitors.  (This is probably not on purpose, since the last stable version predates Retina (I think), and definitely predates Apple's decision to deliberately make things look worse than they used to on non-Retina machines.)

**Comment by chdiza** [comment]
OK, to sum up the current state of play (because this Issue is starting to get complicated):

(1) I noticed two things about the new ANSI coloring facility:  (a) it colors the wrong number of characters and the wrong particular characters under certain circumstances.  (b) the styled text that results looks overantialised.

  - Item (a) is currently being hacked on by @asmagill.

  - Item (b) seems to go away, or at least be lessened, if the background is set to be non-transparent; the fact that it looks bad when transparent is probably just a fact about OS X and not HS.  But we must also be on the lookout for Retina-optimization inadvertently ruining the non-Retina experience by causing overantialiasing, since this happens in some other apps.

(2) In the process of testing the new ANSI coloring facility by trying to replicate some of my Geeklets, I stumbled upon some kind of memory leak or leakish thing.  This is almost certainly an issue with `hs.timer` and not with the new ANSI stuff.

**Comment by asmagill** [comment]
@chdzia, WRT to memory, modifying your sample so that the function looks like this:

~~~lua
function refresh_my_ls()
  if not my_ls then
    my_ls = hs.drawing.text(my_ls_spot, "")
    my_ls:sendToBack()
    my_ls:show()
  end
  my_ls:setText(run_my_ls())
end
~~~

seems to be better behaved on my machine wrt to memory... your code destroyed and recreated a large drawing item every time it was run; this version doesn't.

Try this out and let me know how it fares on your setup

**Comment by chdiza** [comment]
> your code destroyed and recreated a large drawing item every time it was run

Because my drawings display the output of a shell script, I found that if I didn't destroy them and create them anew, I had no way to refresh the script output.  I could only create duplicate drawings that landed on top of each other and looked weird.  Thus I started detroying the prior incarnations.

I could swear I tried `:setText()` as you have in the last line of the function, but that it didn't work.  I'll try again soon when I get a chance.

But that aside: there still shouldn't have been any memory leaking, right?  If a drawing is destroyed, its memory should be given up.  And were those drawings actually "large"?  It didn't seem like any of that should require even 1MB of memory, much less 4 or 8.

**Comment by asmagill** [comment]
poor choice of words... I didn't mean large in the sense of memory (I honestly don't know how much it should take up given that its mostly empty), but large in the sense that 900x900 is 2/3 of my laptops primary display

I would agree that ideally, it shouldn't show memory growth with paired new and deletes, but if you confirm that making this change does in fact dampen down the memory growth, it suggests that we need to look closer at how __gc is working for hs.drawing in general.

**Comment by chdiza** [comment]
I just did a quick test.  The leak seems to not be there.  The new pattern (which repeats, I believe, with every firing of the timer) is: up 4MB, then down 4MB, then up 4, then down 4.

So at least with this test init file, it's not ballooning.  I'll put it to more tests.

Out of curiosity, in your revamped function, why did you make this part:
```lua
my_ls = hs.drawing.text(my_ls_spot, "")
```
have the emtpy string as second argument?  Or, better phrased, my question is: is there anything wrong or suboptimal about doing it like the below?
```lua
function refresh_my_ls()
  if not my_ls then
    my_ls = hs.drawing.text(my_ls_spot, run_my_ls())
    my_ls:sendToBack()
    my_ls:show()
  else
    my_ls:setText(run_my_ls())
  end
end
```

**Comment by asmagill** [comment]
Stylistic choice... I was thinking in terms of not having an else clause.

**Comment by cmsj** [comment]
I think it's worth noting that just because RAM usage is increasing, doesn't mean it won't decrease later as the objects are garbage collected. Lua's defaults for garbage collection are not super aggressive, and are (partially) a function of how much memory is being consumed *in Lua*. Our objects may be large on the ObjC side, but are typically very small from Lua's perspective, so it could take a long time for them to be collected.

It may be worth us running some tests to see if we should make the collector more aggressive by default.

**Comment by cmsj** [comment]
I've just quickly thrown together a couple of functions in hs.crash that will help with automated reporting of memory usage, and attempt to wrap up a forced GC cycle (although so far I've yet to see the process' resident size decrease as a result of a forced GC run, but I've not tried to stress it very hard with lots of objects)

**Comment by chdiza** [comment]
>  just because RAM usage is increasing, doesn't mean it won't decrease later as the objects are garbage collected

When the plus-8MB/minus-4MB pattern was going on, HS would get up to 800MB of RAM.  That seems awfully huge.  (Plus, I never saw it go back down to normal levels---I would quit it to keep it from ballooning further.)  All of which is just to say, I agree that the collector needs investigation :)

A possibly related issue in testing the new ANSI facility's ability to simulate geeklets appeared.  In the above sample init file, if the script executed is `ruby --version`, the RAM used by HS is about 10MB higher than if the script run is `ls`.  And this 10MB goes unreclaimed as far as I can tell.  I know that ruby will command more RAM than ls, but I don't understand why---if indeed the extra 10MB is for ruby---it doesn't get released.

**Comment by cmsj** [comment]
@chdiza I agree that 800MB is unreasonably huge, but from Lua's perspective, it would have been a few tens of KB of userdata pointers, so it wouldn't have been rushing to do much GC.

When you're talking about the memory usage of Hammerspoon, which column in Activity Monitor are you referring to? The only one that's really worth looking at is "Real Mem", which is how much physical RAM/swap the process is using.

Also also, I tend to find that running the non-release build of Hammerspoon from Xcode, uses way more RAM than using the release builds on their own.

tl;dr, this stuff is super subtle and non-trivial ;(

**Comment by chdiza** [comment]
> which column in Activity Monitor are you referring to? The only one that's really worth looking at is "Real Mem", which is how much physical RAM/swap the process is using.

I was looking at "Memory", which looks to me like it's nearly always *smaller* than what "Real Mem" says.  I'll conduct my tests again while looking at "Real Mem".

> I tend to find that running the non-release build of Hammerspoon from Xcode, uses way more RAM than using the release builds on their own.

How can I build a release build?  Maybe I can't, because of code-signing?

**Comment by cmsj** [comment]
It's a while since I tried to do a release build without the code signing stuff available, but I think it should work. Up near the top left of Xcode you should see it say "Hammerspoon > My Mac" - click on the Hammerspoon part of that and choose "Release". Typically I would then navigate down the folder structure on the left to Hammerspoon -> Products, right click on the Hammerspoon.app, reveal it in Finder and double click it, just to be sure that Xcode isn't involved at all.

**Comment by chdiza** [comment]
Now with bf4ccade, the crashes seem to have disappeared, which means I can better monitor the memory-leak issue.

What seems to be the case now is that it is still leaking but much more slowly.  For instance, when I fire up HS fresh, and after waiting for my "geeklets" to display, Activity Monitor shows it uses about 22MB.  Within maybe 90 minutes or so of being untouched (not hs.reload()ed), it's up to 65MB.  It seems to get there at a relatively constant rate of increase.

It's possible that eventually HS would garbage collect this back down, so I'll have to keep an eye on it for longer time-periods.

**Comment by chdiza** [comment]
> BTW I do have a fix for the unicode multi-byte issue you pointed out, but its in the Lua code

@asmagill is this fix in a place I can access for testing?

**Comment by asmagill** [comment]
The obj-c version is in pull #625... the lua version for comparison, if you care, is commented out in the init.lua file, but I threw a couple of tests with about 200 unicode chars of varying byte lengths at the compiled version and it worked fine for all of them.  The pull also has the hs.drawing:wantsLayer(bool) method (currently not in the docs, though... need to fix that) it defaults to false, but try setting it to true and see how that affects the anti-aliasing on your monitor mentioned earlier -- it does help noticeably on mine.

**Comment by chdiza** [comment]
@asmagill OK, I tried out #625 and the unicode multi-byte ANSI choking seems to be gone!  Thanks!

I'll try out the `wantsLayer` a bit later.

**Comment by chdiza** [comment]
> The pull also has the hs.drawing:wantsLayer(bool) 

I tried this.  It made text whose background was transparent look slightly better (though more faded).  It made text that didn't have a transparent background extremely faint.

**Comment by chdiza** [comment]
I want to express a huge thank you to everyone who worked on hs.styledtext, hs.drawing, hs.task, etc. to make HS more usable in a GeekToolish way.  You guys rock :smile:

I've been using HS as a complete GeekTool replacement for nearly a month, and (at least for my particular set of geeklets), there's nothing I could do in GeekTool that I can't do in HS.

There are currently only two respects in which GeekTool is still superior to HS.  (They are relatively minor, especially the first one).  They are:

1. GeekTool's "Image Geeklets" ask for a path for the image that's to be displayed.  This can be either a local path, or a remote URL.  In HS, hs.image can't be fed a remote URL to an image.  So for my "fake" Image Geeklets, I end up using hs.task to call `curl` to download the remote image, and *then* use hs.image to show it.  (I also tried using hs.webview, which worked, but found this to be rather more RAM intensive.)  In a perfect FakeGeekTool world, hs.image (or some other special thing like "hs.remoteImage") would be able to take a remote URL to an image file and show it.

2. HS still seems to leak memory, even if not to the same degree as described earlier in this thread.  I have a hunch that the source of the leak is that either hs.styledtext objects  don't get garbage-collected, or the hs.task objects that call the shell-geeklet scripts aren't getting collected.  (If I stop the hs.timers that refresh the geeklets at regular intervals, I don't seem to see the leak continue to grow, but RAM usage also never drops back down.)  Even when I leave "FakeGeekTool" up all day long, or for 48 straight hours, there seems to be no eventual collection.

But as I said, these are minor.  And of course, there are many respects in which fake GeekTool is superior to real GeekTool :)

**Comment by asmagill** [comment]
I'd be the last one to deny the possibility that styledtext could be leaking... I certainly don't want it to, but it was not as simple a module as I had hoped when I started it!  I'm guessing that you probably have a number of timers running with "Geeklets" and NSTasks triggering callbacks, so it's a safe bet that the lua state is pretty active (i.e. not idle)...  I'd be curious what making the garbage collector more aggressive does to your memory usage... if I read section 2.5 of the Lua 5.3 manual correctly, I *think* adding `collectgarbage("setpause", 100)` will cause garbage collection to happen every time the lua state is active -- the default of 200 waits until memory use has doubled.  Not sure what impact it will have on performance, so maybe a value of 150 is a better staring point.

At any rate, I am planning to review styledtext and webview again soon and see if anything jumps out... I needed a break and to spend time on other things to "reset", so to speak.

**Comment by chdiza** [comment]
Is using `collectgarbage()` something I can (informatively) test myself in the hs.console?  If so, I'll play with it and report back any effects it seems to have.

I do indeed have a number of timers running, a couple of which fire every 5 seconds.

**Comment by asmagill** [comment]
Yes, it's a built-in lua command; you can type it in the console or put it into your init.lua, whichever makes the most sense for your setup (I restart HS a lot, so I'd probably put it into my init.lua so I didn't have to remember to keep typing it in, but if you run for a long time between restart/reloads, whichever makes the most sense to you.

**Comment by chdiza** [comment]
Closing, as the original question was answered and the later memory leak appears to be gone.

Many thanks!

**Comment by asmagill** [comment]
Forgot about the `imageFromURL` idea... I'll add it to the webview updates in #920, so expect to see it soon.

**Comment by chdiza** [comment]
> Forgot about the imageFromURL idea... I'll add it to the webview updates in #920, so expect to see it soon.

@asmagill Thanks for that nice addition.  It works, and I think it fills a needed role in HS.

It turns out that I won't be using it in fake GeekTool, however.  That's because everything in fake GeekTool gets blocked while the `imageFromURL` is fetching the image.  I have three fake ImageGeeklets that would use this.  If I use imageFromURL in them, all of HS (and thus fake GeekTool) is blocked until the three images are fetched.  They aren't big images, but sometimes the network is not super-fast or the image server is not super-fast.

Prior to the existence of imageFromURL, my workaround was to make each fake ImageGeeklet use hs.task to call a shell script that uses `curl` to fetch the images, then used plain old hs.drawing.image() on the local paths of the curl'ed pngs.  It's hacky, but because it uses hs.task, the image fetching can happen in the background while other fake Geeklets can load and display.

**Comment by asmagill** [comment]
@chdiza, should we add a "fetch image in the background" function that takes a callback that receives the `hs.image` object? I don't think it would be that hard to do (suspect it will be similar to what we do with `hs.wifi` and background scanning, though I haven't really looked into it yet).

**Comment by chdiza** [comment]
>  should we add a "fetch image in the background" function that takes a callback that receives the hs.image object?

That would probably be good.  (I can easily imagine that someone wants HS to put a large image, from a URL, on the screen, and doesn't want HS blocking while it downloads.  Or a not-so-large image over a slow network.  Or lots of images, which would otherwise be fetched in series.)

Don't knock yourself out over this, though.  I've had a fine workaround in place for months, and anyone in the imaginary situations above can do something similar.

**Comment by heptal** [comment]
Yea, you could always define it yourself easily with hs.http's async functions. 

@asmagill NSImage's initWithContentsOfURL might be replaced by initByReferencingURL which creates a lazy reference to the remote resource? Not sure if it blocks when it finally fetches the data, though.

**Comment by asmagill** [comment]
@heptal, tried that... it just changes when the blocking occurs.

Re the http async functions, I've been pondering that... I've been mulling over adding some things to image and pasteboard that could work with images as raw data... my original purpose was to be able to save images in the application defaults for use with a clipboard manager, but it occurred to me that if we have a way to create an image from raw data, it offers a Hammerspoon based solution to this problem as well.

**Comment by NightMachinary** [comment]
@chdiza Can you share some screenshots of your geeklets?

**Comment by chdiza** [comment]
Sure, once I've anonymized their output.  I'm afraid they're not fancy, just text.  They're meant to correspond to GeekTool's "shell geeklets".
